### ISSUE TITLE
SL.dbarts modification

### ISSUE THREAD

**Issue by benkeser** [ecpolley/SuperLearner#102]
Hi all,
Was fooling around with `SL.dbarts` and noticed that it sometimes returns values outside (0,1) when Y is binary. Seems like `bart` wants a factor input for the outcome variable if it's to properly handle (0,1) outcomes. Here's a simple modification that does the trick.
```{r}
SL.dbarts = function(Y, X, newX, family, obsWeights, id,
                     sigest = NA,
                     sigdf = 3,
                     sigquant = 0.90,
                     k = 2.0,
                     power = 2.0,
                     base = 0.95,
                     binaryOffset = 0.0,
                     ntree = 200,
                     ndpost = 1000,
                     nskip = 100,
                     printevery = 100,
                     keepevery = 1,
                     keeptrainfits = T,
                     usequants = F,
                     numcut = 100,
                     printcutoffs = 0,
                     nthread = 1,
                     keepcall = T,
                     verbose = F,
                     ...) {

  .SL.require("dbarts")
if(family$family == "binomial"){
  y.train <- factor(Y)
}else{
  y.train <- Y
}
  model =
    dbarts::bart(x.train = X,
                 y.train = y.train,
                 # We need to pass newX in directly due to lack of prediction.
                 x.test = newX,
                 sigest = sigest,
                 sigdf = sigdf,
                 sigquant = sigquant,
                 k = k,
                 power = power,
                 base = base,
                 binaryOffset = binaryOffset,
                 weights = obsWeights,
                 ntree = ntree,
                 ndpost = ndpost,
                 nskip = nskip,
                 printevery = printevery,
                 keepevery = keepevery,
                 keeptrainfits = keeptrainfits,
                 usequants = usequants,
                 numcut = numcut,
                 printcutoffs = printcutoffs,
                 nthread = nthread,
                 keepcall = keepcall,
                 verbose = verbose)

  # TODO: there is no predict!
  #pred = predict(model, newdata = newX)
  if (family$family == "gaussian") {
    pred = model$yhat.test.mean
  } else {
    # No mean is provided for binary Y :/
    pred = colMeans(model$yhat.test)
  }

  fit = list(object = model)
  class(fit) = c("SL.dbarts")
  out = list(pred = pred, fit = fit)
  return(out)
}
```

**Comment by benkeser** [comment]
Just kidding, now seeing that bart claims it fits probit link if all Y are in 0,1. But it still returns values outside (0,1). Seems like a `bart` problem, not a `SuperLearner` problem. Closing the issue...

**Comment by ck37** [comment]
well BartMachine definitely has this problem - I have a fix in my ck37r package (BartMachine2) but haven't had time to do a pr yet.

Just curious, are you using dbarts because bartMachine is a pain to get working, or for other reasons?

-ck.mobile

**Comment by benkeser** [comment]
Alright, sorry for being obnoxious and spamming here, but just noticed in the `bart` documentation that if binary outcomes are used then the output of `bart` needs to be `pnorm`'ed to get back to the probability scale. Thus, this modification should work:
```
SL.dbarts = function(Y, X, newX, family, obsWeights, id,
                     sigest = NA,
                     sigdf = 3,
                     sigquant = 0.90,
                     k = 2.0,
                     power = 2.0,
                     base = 0.95,
                     binaryOffset = 0.0,
                     ntree = 200,
                     ndpost = 1000,
                     nskip = 100,
                     printevery = 100,
                     keepevery = 1,
                     keeptrainfits = T,
                     usequants = F,
                     numcut = 100,
                     printcutoffs = 0,
                     nthread = 1,
                     keepcall = T,
                     verbose = F,
                     ...) {

  .SL.require("dbarts")

  model =
    dbarts::bart(x.train = X,
                 y.train = Y,
                 # We need to pass newX in directly due to lack of prediction.
                 x.test = newX,
                 sigest = sigest,
                 sigdf = sigdf,
                 sigquant = sigquant,
                 k = k,
                 power = power,
                 base = base,
                 binaryOffset = binaryOffset,
                 weights = obsWeights,
                 ntree = ntree,
                 ndpost = ndpost,
                 nskip = nskip,
                 printevery = printevery,
                 keepevery = keepevery,
                 keeptrainfits = keeptrainfits,
                 usequants = usequants,
                 numcut = numcut,
                 printcutoffs = printcutoffs,
                 nthread = nthread,
                 keepcall = keepcall,
                 verbose = verbose)

  # TODO: there is no predict!
  #pred = predict(model, newdata = newX)
  if (family$family == "gaussian") {
    pred = model$yhat.test.mean
  } else {
    # No mean is provided for binary Y :/
    pred = colMeans(pnorm(model$yhat.test))
  }

  fit = list(object = model)
  class(fit) = c("SL.dbarts")
  out = list(pred = pred, fit = fit)
  return(out)
}
```

**Comment by benkeser** [comment]
I'd tried `bartMachine` in the `caret` package before, but found it to be really slow. Too slow to want to include in a super learner at least. I admit that I spent 0 time looking at the documentation to figure out if the speed was my fault or inherent to the method. Are there performance benefits to `bartMachine` over `dbarts`?

**Comment by ck37** [comment]
Well my main issue with dbarts is that it can't predict on new data - any test data has to be supplied during training.  (https://github.com/vdorie/dbarts/issues/7)

For bartMachine I've found that as long as multiple cores are used (bartMachine::set_bart_machine_num_cores(x))
 it isn't too slow, but I should try dbarts again and compare. Avoiding rJava is a big advantage to dbarts.

**Comment by benkeser** [comment]
Yeah, that is a bit annoying with `dbarts`. Good to know about parallel `bartMachine`.

**Comment by ecpolley** [comment]
Thanks, is the only change the line for prediction in the binomial case?

       # No mean is provided for binary Y :/
       pred = colMeans(pnorm(model$yhat.test))

**Comment by benkeser** [comment]
Yep

**Comment by ecpolley** [comment]
done

**Comment by ck37** [comment]
It would be good to add a test for this too.

**Comment by ecpolley** [comment]
Test if predicted values are outside [0, 1]?

**Comment by ck37** [comment]
Yup, ideally (but not mandatory) with a test case that would cause that to
happen if we used the original code.

**Comment by ecpolley** [comment]
I was thinking we could add a warning in the method.* when family is binomial to check the Z matrix for values outside [0, 1].

**Comment by ck37** [comment]
Yeah that would be great too.

**Comment by benkeser** [comment]
Hey guys, one more note on `SL.dbarts` that I just ran into. The `bart` function doesn't play nicely when `X` comes in as a `data.frame` with a single column. I submitted [an issue](https://github.com/vdorie/dbarts/issues/8) to the `dbarts` maintainer, but for now I think this little blip can be added into `SL.dbarts`  before the call to `bart` to fix the error. 

```r
  if(dim(X)[2] == 1){
    X <- X[,1]; newX <- newX[,1]
  }
```
